feat(engine): Add tiered group system for scalable NPC group access

puncar-dev opened the pull request:
## Summary

- **Implement 3-tier NPC group system**: Inner Circle (12 max), Community (50 max), Followers (500 max) per NPC
- **Scale capacity**: From 1,680 total users to 78,680 (46x increase for 140 NPCs)
- **Asymmetric information mechanics**: Tier-specific alpha sharing (full → partial → public)

## Type

- [x] Feature

## Context / Links

- Addresses scalability for 300,000 users
- Builds on existing NPC group dynamics

## Scope

- [x] Domain / game engine (`packages/engine`)
- [x] DB (`packages/db`)
- [x] API infra (`packages/api`)
- [x] Web API routes (`apps/web`)
- [x] Tests (`packages/testing`)
- [x] Docs

## Changes

### New Files
- `packages/engine/src/services/tier-config.ts` - TIER_CONFIG constants and helpers
- `packages/engine/src/services/tiered-group-service.ts` - Core service
- `packages/engine/src/prompts/game/group-messages-tier{1,2,3}.ts` - Tier-specific prompts
- `packages/db/drizzle/migrations/0011_add_tiered_group_system.sql` - Migration
- `packages/db/drizzle/migrations/0012_rollback_tiered_group_system.sql` - Rollback
- `apps/web/src/app/api/tiers/` - User tier status endpoints
- `apps/web/src/app/api/admin/tiers/` - Admin analytics endpoint
- `packages/testing/integration/tiered-group-system.test.ts` - 63 tests

### Modified Files
- `npc-group-dynamics-service.ts` - Integrated tiered invitations and promotions
- `notification-service.ts` - Added tier promotion/demotion notifications
- `ChatListItem.tsx` - Added tier badges (Gold/Blue/Gray)
- Chat API routes - Return tier info

## Review guide

**Start here**: `packages/engine/src/services/tier-config.ts` for tier definitions, then `tiered-group-service.ts` for core logic.

**Risk level**: Medium - DB migration required

## Test plan

- [x] `bun run check` (Biome format)
- [x] `bun run lint` (our files pass, pre-existing errors in staging)
- [x] `bun test packages/testing/integration/tiered-group-system.test.ts` (63 pass)
- [ ] `bun run typecheck` (pre-existing errors in babylon-typescript-agent-example)

## Ops / Migration / Deployment

- **DB migration required**: `bun run db:migrate` to apply 0011
- **Rollback available**: 0012_rollback_tiered_group_system.sql
- **No env changes**

## Breaking changes

None - existing groups continue to work, will be migrated to Tier 1 by migration

## Security / privacy

No impact

## Screenshots

No UI changes beyond tier badges in chat list (Gold shield = Inner Circle, Blue = Community, Gray = Followers)

## Checklist (author)

- [x] Self-review done
- [x] Base branch correct (`staging`)
- [x] No screenshots needed (minor badge addition)
- [x] Handlers thin and portable
- [x] Domain logic in packages
- [x] Tests added (63 tests covering config, helpers, and service methods)
- [x] Code owners requested

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Tiered group system with three engagement levels (Inner Circle, Community, Followers).
  * Automated invites, promotions, and demotions based on engagement and activity rules.
  * Tier-specific membership limits, messaging frequency, and guidance for richer NPC interactions.
  * New user-facing views: per-NPC tier listings, individual tier status, and global tier analytics/capacity insights.
  * Existing NPC groups and memberships initialized into Tier 1 (Inner Circle) to preserve continuity.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

slkzgm commented on packages/engine/src/services/tiered-group-service.ts:
```diff
@@ -0,0 +1,1085 @@
+    return {
+      userId,
+      npcId,
+      currentTier: membership?.tier as TierLevel | null,
```
Potential runtime shape issue: `currentTier: membership?.tier as TierLevel | null` can become `undefined` when `membership` is missing. Would you switch to something like `currentTier: (membership?.tier ?? null) as TierLevel | null` (and maybe validate 1/2/3) so the API always returns `null` when the user is not in any tier?

slkzgm commented on packages/engine/src/services/tiered-group-service.ts:
```diff
@@ -0,0 +1,1085 @@
+      };
+    }
+
+    // Check cooldown
```
Question on cooldown semantics: this check looks at any `GroupChatMembership` joined in the last `INVITE_COOLDOWN_HOURS` regardless of `isActive` (and regardless of NPC). In staging, the cooldown logic seemed based on latest active membership. Is the stricter global cooldown intentional? If not, consider filtering by `isActive = true` (and/or by `npcAdminId`).

slkzgm commented on packages/engine/src/services/tiered-group-service.ts:
```diff
@@ -0,0 +1,1085 @@
+    }
+
+    // Add to chat participants
+    await db.insert(chatParticipants).values({
```
Requesting clarification: `ChatParticipant` is unique on `(chatId, userId)` and `GroupChatMembership` is unique on `(userId, chatId)`. If a user ever gets removed from a tier and later re-enters the same tier (e.g., Tier 2 -> Tier 3 -> Tier 2), these inserts will likely conflict unless we reactivate existing rows. Would an `onConflictDoUpdate` (or an update-then-insert fallback) be safer here? Also worth wrapping both inserts in a transaction.

slkzgm commented on packages/engine/src/services/tiered-group-service.ts:
```diff
@@ -0,0 +1,1085 @@
+      return null;
+    }
+
+    // Remove from current tier
```
Promotion path updates 2 tables then inserts 2 rows. Without a transaction, a mid-way failure can leave the user removed from the old tier but not added to the new one. Also, the inserts into `targetTierInfo.chatId` will conflict if the user was previously in that tier chat (unique constraints). Can we make this atomic and re-activate existing participant/membership rows instead of inserting unconditionally?

slkzgm commented on packages/engine/src/services/tiered-group-service.ts:
```diff
@@ -0,0 +1,1085 @@
+
+    const lowerTier = getLowerTier(status.currentTier);
+
+    // Remove from current tier
```
Similar concern on demotion: we deactivate the current tier before confirming the target tier exists/has space, and then we insert into the lower tier. If the lower-tier inserts fail (or the tier chat is missing), the user ends up in no tier. Would you consider (1) checking target tier availability first, (2) wrapping the whole move in a transaction, and (3) reactivating existing participant/membership rows for the lower tier to avoid unique constraint conflicts?

slkzgm commented on packages/engine/src/services/tiered-group-service.ts:
```diff
@@ -0,0 +1,1085 @@
+      return { invites, errors };
+    }
+
+    for (const actor of actors) {
```
This runs for every actor on every tick and calls `ensureAllTiersExist` + `getEligibleUsersForNpc` (which computes engagement scores). At 100+ NPCs this could get expensive fast. Is this intentionally per-tick, or should invites run at a lower frequency / for a subset of NPCs, or use cached/precomputed engagement rankings?

slkzgm commented on packages/engine/src/services/tiered-group-service.ts:
```diff
@@ -0,0 +1,1085 @@
+          isNull(groupChatMemberships.id) // Not in any group with this NPC
+        )
+      )
+      .limit(50); // Reduce to 50 since we calculate engagement for each
```
`limit(50)` without an explicit order makes the candidate pool effectively arbitrary (and not truly "top engaged" users). If the intent is sampling, maybe add a deterministic/random order; if the intent is "top engaged", we probably need a cheaper prefilter or a precomputed score table/index so we do not compute engagement for a mostly random slice.

slkzgm commented on packages/db/drizzle/migrations/0011_add_tiered_group_system.sql:
```diff
@@ -0,0 +1,38 @@
+  AND "tier" IS NULL;
+
+-- Update existing memberships to Tier 1
+UPDATE "GroupChatMembership"
```
Migration question: `UPDATE "GroupChatMembership" SET "tier" = 1 WHERE "tier" IS NULL;` is very broad. Are we 100% sure `GroupChatMembership` only tracks NPC groups? If not, safer would be to join `Chat` and scope to chats that were migrated to tiered NPC groups (e.g., `Chat.isGroup = true AND Chat.npcAdminId IS NOT NULL`).

slkzgm commented on apps/web/src/app/api/chats/route.ts:
```diff
@@ -425,6 +425,11 @@ export const GET = withErrorHandling(async (request: NextRequest) => {
           qualityScore: membership.qualityScore,
           lastMessageAt: membership.lastMessageAt,
           updatedAt: chat.updatedAt,
+          // Tier information
```
Nice adding tier metadata here. One consistency question: the `?all=true` public branch still returns chats without `tier/tierName/maxMembers/npcAdminId` (CodeRabbit flagged this too). If the discovery UI consumes these fields, should we include them there as well (even if `null`)?

slkzgm commented on apps/web/src/components/chats/ChatListItem.tsx:
```diff
@@ -39,8 +71,31 @@ export function ChatListItem({
+            <div
+              className={cn(
+                'chat-button flex h-10 w-10 shrink-0 items-center justify-center rounded-full',
+                tierBadge ? 'bg-sidebar-accent/50' : 'bg-sidebar-accent/50'
```
Nit: `tierBadge ? 'bg-sidebar-accent/50' : 'bg-sidebar-accent/50'` is a no-op. If there is no intended visual difference, we can drop the conditional; if there is, we probably want different classes.

slkzgm commented on apps/web/src/components/chats/ChatListItem.tsx:
```diff
@@ -39,8 +71,31 @@ export function ChatListItem({
+            >
+              {tierBadge ? (
+                <tierBadge.Icon
+                  className={cn('h-5 w-5', tierBadge.className.split(' ')[1])}
```
Minor robustness: deriving icon color via `tierBadge.className.split(' ')[1]` is brittle (depends on class ordering). Could we store an explicit `iconClassName` (or use `text-*` separately) to avoid accidental breakage?

slkzgm commented on packages/api/src/services/notification-service.ts:
```diff
@@ -638,3 +638,87 @@ export async function notifyGroupChatMessage(
+  fromTier: number,
+  toTier: number
+): Promise<void> {
+  const result = await db
```
Suggestion: both functions query the NPC name on every call and repeat the tier-name mapping. If promotions/demotions happen in batch, this is N queries. Could we pass `npcName` from the caller (or prefetch once per NPC) and reuse `TIER_CONFIG`/a shared helper for tier display names to avoid drift?

slkzgm commented on apps/web/src/app/api/tiers/[npcId]/route.ts:
```diff
@@ -0,0 +1,33 @@
+    request: NextRequest,
+    context: { params: Promise<{ npcId: string }> }
+  ) => {
+    const { npcId } = await context.params;
```
Should we validate `npcId` upfront (e.g., ensure it is a known actor) and return 404/400? Right now we compute engagement and return empty tiers for unknown ids, which may be fine but I want to confirm expected behavior.

slkzgm reviewed (changes requested):
Thanks for the thorough PR + docs/tests — the overall structure (tier config + service layer + thin routes) is heading in a good direction.

## Requested changes (blockers)

1) **Tier transitions vs unique constraints**
I believe tier transitions will hit unique constraints in real promotion/demotion cycles:
- `ChatParticipant` is unique on `(chatId, userId)`
- `GroupChatMembership` is unique on `(userId, chatId)`

But `inviteUserToTier` / `promoteUser` / `demoteUser` mark the current rows inactive and then unconditionally `insert` into the target tier chat. A simple Tier 3 → Tier 2 → Tier 3 path would attempt to insert into the Tier 3 chat again while the old row still exists (inactive), which should fail.

2) **Atomicity for promotions/demotions**
Promotions/demotions are multi-table moves (update membership + update participants + insert participants + insert membership). Without a transaction, any mid-way failure can leave users in inconsistent states.

3) **`getUserTierStatus` response shape**
`currentTier: membership?.tier as TierLevel | null` can become `undefined` when the user has no membership; I think we should normalize to `null`.

## Major / follow-ups

- **Per-tick performance**: `processTickInvitations` runs `ensureAllTiersExist` + engagement score computation for each actor. I would like to confirm tick frequency and scale assumptions (this could become expensive at 100+ NPCs).
- **API shape consistency**: `GET /api/chats?all=true` still returns public chats without tier metadata (not in this diff hunk, but it may matter for chat discovery UI).

## Minor

- UI tier badge implementation has a couple small cleanups (redundant ternary, class split fragility).
- Tier promotion/demotion notifications duplicate tier name mapping and do an NPC lookup per call (could be optimized / de-duplicated).

**Verdict: Request changes** (mainly due to unique-constraint + atomicity concerns).

SYMBaiEX committed:
feat(engine): add tiered group system for unified Group schema

SUMMARY:
- Implement 3-tier NPC group system aligned with unified Group/GroupMember schema
- Tier 1 (Inner Circle): 12 members, full alpha, 25% message frequency
- Tier 2 (Community): 50 members, partial alpha, 15% message frequency
- Tier 3 (Followers): 500 members, public content, 5% message frequency
- Total capacity: 140 NPCs × 562 slots = 78,680 users (46x increase)

CHANGES:
- packages/db: Add tier, maxMembers, parentGroupId to Group table
- packages/db: Add tier, promotedAt, demotedAt, previousTier to GroupMember
- packages/db: Migration 0012_add_tiered_group_system.sql
- packages/engine: New tier-config.ts with TIER_CONFIG constants
- packages/engine: New TieredGroupService for tier management
- packages/engine: Integrated tier-based messaging in NPCGroupDynamicsService

SCALE:
- Previous: 12 users per NPC (1,680 total for 140 NPCs)
- New: 562 users per NPC (78,680 total capacity)

SYMBaiEX committed:
feat(engine): enhance tiered group system with validation and messaging improvements

SUMMARY:
- Introduce validation for tier levels and improve messaging guidance for NPC group chats.
- Implement distributed locking in tier invite and promotion processes to prevent race conditions.
- Optimize batch queries for promotions and demotions to enhance performance.

CHANGES:
- packages/db: Added composite indexes for group member queries.
- packages/engine: Refactored NPCGroupDynamicsService to utilize new tier messaging guidance.
- packages/engine: Enhanced TieredGroupService with validation checks for tier levels and improved transaction handling.
- packages/engine: Introduced new utility functions for tier validation and messaging guidance.

IMPACT:
- Improved maintainability and performance of the tiered group system, ensuring robust handling of user promotions and demotions.

SYMBaiEX committed:
refactor(engine): streamline tier validation and improve error messaging

SUMMARY:
- Simplified the implementation of the `isValidTier` function for better readability.
- Enhanced error messaging in `assertTierLevel` to maintain clarity and consistency.
- Cleaned up code formatting in the `TieredGroupService` to improve maintainability.

CHANGES:
- packages/engine/src/services/tier-config.ts: Refactored type guard and assertion functions for tier validation.
- packages/engine/src/services/tiered-group-service.ts: Improved logging and transaction handling for tier demotion processes.

IMPACT:
- Improved code clarity and maintainability, ensuring robust tier validation and error handling in the tiered group system.

SYMBaiEX committed:
refactor(engine): update tier level handling and improve messaging guidance

SUMMARY:
- Removed the local definition of `TierLevel` in favor of the canonical definition from `@babylon/shared`.
- Enhanced the `TieredGroupService` to batch generate IDs for new tier groups, improving efficiency.
- Optimized the `getNpcTiers` method to use a single query with joins, reducing the N+1 query problem.

CHANGES:
- packages/db/src/schema/messaging.ts: Updated to reference `TierLevel` from shared types.
- packages/engine/src/services/tier-config.ts: Re-exported `AlphaLevel` and `TierLevel` for convenience.
- packages/engine/src/services/tiered-group-service.ts: Improved ID generation and logging for tier operations.

IMPACT:
- Streamlined tier management processes, enhancing performance and maintainability of the tiered group system.

SYMBaiEX committed:
refactor(engine): clean up imports and improve code formatting in tier services

SUMMARY:
- Adjusted import order in `tier-config.ts` for consistency.
- Reformatted `groupBy` method in `tiered-group-service.ts` for improved readability.
- Enhanced transaction handling in tier demotion process for clarity.

CHANGES:
- packages/engine/src/services/tier-config.ts: Cleaned up import statements.
- packages/engine/src/services/tiered-group-service.ts: Reformatted method calls and improved code structure.

IMPACT:
- Improved code clarity and maintainability in tier management services, facilitating easier future modifications.

SYMBaiEX committed:
refactor(engine): optimize database queries and update Claude model usage

SUMMARY:
- Updated Claude model to use Opus 4.5 for improved code review quality in workflows.
- Optimized database queries in `npc-group-dynamics-service.ts` and `tiered-group-service.ts` to reduce N+1 query issues by utilizing JOINs.
- Enhanced the `ensureAllTiersExist` method to batch query existing tiers, improving efficiency.

CHANGES:
- .github/workflows/claude-code-review.yml: Updated Claude model and documentation links.
- .github/workflows/claude.yml: Adjusted Claude model for better understanding and implementation.
- packages/engine/src/services/npc-group-dynamics-service.ts: Refactored to use a single query with LEFT JOIN for tier data.
- packages/engine/src/services/tiered-group-service.ts: Improved batch querying for existing tiers and optimized transaction handling.

IMPACT:
- Enhanced performance and maintainability of the tiered group system, ensuring efficient data retrieval and improved code review processes.

SYMBaiEX committed:
refactor(engine): improve code formatting in tiered group service

SUMMARY:
- Reformatted the assignment of `participantId` in `TieredGroupService` for better readability.

IMPACT:
- Enhanced code clarity, contributing to improved maintainability in the tiered group system.
